File: week2/assign2.py
def function1(name):
    people: dict[str, tuple[int, int]]={
        "悟空": (0, 0),
        "辛巴": (-3, 3),
        "貝吉塔": (-4, -1),
        "特南克斯": (1, -2),
        "佛利沙": (4, -1),
        "丁滿": (-1, 4)
    }
    section1 = ["悟空", "辛巴", "貝吉塔", "特南克斯"]
    section2 = ["佛利沙", "丁滿"]

    distance: dict[str, int] = {}
    
    def find_section(name):
        if name in section1:
            return 1
        elif name in section2:
            return 2
    target_section = find_section(name)
    extra_point = 2

    for n in people:
        if n == name:
            continue
        else:
            diff = abs(people[n][0] - people[name][0]) + abs(people[n][1] - people[name][1])
        if find_section(n) != target_section:
            diff = diff + extra_point
        distance[n] = diff
        
    value_max = max(distance.values())
    value_min = min(distance.values())
    max_key = []
    min_key = []
    for key, value in distance.items():
        if value == value_max:
            max_key.append(key)
        if value == value_min:
            min_key.append(key)
    print("最遠", "、".join(max_key), ";最近", "、".join(min_key))


print("================Task1================")
function1("辛巴")
function1("悟空")
function1("佛利沙")
function1("特南克斯")

def func2(ss, start, end, criteria):
    # divide each element in criteria,
    def divide(criteria):
        criteria = criteria.replace(" ", "")
        for op in ("<=", ">=", "="):
            if op in criteria:
                field, raw = criteria.split(op, 1)
                if field == "name":
                    if op != "=":
                        raise ValueError("name can only use '='")
                    return field, op, raw
                elif field == "r":
                    return field, op, float(raw)
                elif field == "c":
                    return field, op, int(raw)
                else:
                    raise ValueError("Error")

    field, op, val = divide(criteria)
    # select qualified service and put into candidates
    raw_candidates = []
    for s in ss:
        if op == "=":
            if s[field] == val:
                raw_candidates.append(s)
        elif op == "<=":
            if s[field] <= val:
                raw_candidates.append(s)
        elif op == ">=":
            if s[field] >= val:
                raw_candidates.append(s)
    # take the best option
    if len(raw_candidates) == 0:
        print("Error") 
        return

    from operator import itemgetter

    if op == "=":
        candidates = raw_candidates
    elif op == ">=":
        candidates = sorted(raw_candidates, key = itemgetter(field))
    elif op == "<=":
        candidates = sorted(raw_candidates, key = itemgetter(field), reverse = True)

    # make a schedule to record time for each service by list[dict{name, start, end}]
    def overlap(start, end, s):
        if (start < s["end"] and end > s["start"]):
            return True
        return False
    # take each candidate compare to each schedule
    # if candidate and schedule use the same service, check if overlap exsist
    # if overlap exsist change candidate continue(outer loop)
    # if not exsist keep going(inner loop) finish all schedules -> success print result and push
    # if finish all candidates (outer loop) -> failure print sorry

    for candidate in candidates:
        conflict = False
        for schedule in schedules:
            if candidate["name"] == schedule["name"]:
                if overlap(start, end, schedule):
                    conflict = True
                    break
        if conflict:
            continue
        print(candidate["name"])
        schedules.append({"name": candidate["name"], "start": start, "end": end})
        return
    print("Sorry")
    return
    
services = [
    {"name": "S1", "r": 4.5, "c":1000},
    {"name": "S2", "r": 3, "c": 1200},
    {"name": "S3", "r": 3.8, "c": 800}
]
schedules = []

print("================Task2================")
func2(services, 15, 17, "c>=800")
func2(services, 11, 13, "r<=4")
func2(services, 10, 12, "name=S3")
func2(services, 15, 18, "r>=4.5")
func2(services, 16, 18, "r>=4")
func2(services, 13, 17, "name=S1")
func2(services, 8, 9, "c<=1500")

# arr = [25, 23, 20, 21, 23, 21, 18, 19, 21, 19, 16, 17]
def func3(n):
    arr = [None] * (n + 1)

    count1 = -2
    count2 = -3
    count3 = 1
    count4 = 2

    arr[0] = 25
    for i in range(1, n+1):
        if i % 4 == 1:
            arr[i] = arr[i-1] + count1
        if i % 4 == 2:
            arr[i] = arr[i-1] + count2
        if i % 4 == 3:
            arr[i] = arr[i-1] + count3
        if i % 4 == 0:
            arr[i] = arr[i-1] + count4
    print(arr[n])
    
print("================Task3================")
func3(1)
func3(5)
func3(10)
func3(30)

def func4(sp, stat, n):
    # for loop put passage into car(difference), if meet car unavailable continue
    # renew the min and record the ans each loop
    # finish loop print result
    import math
    if len(sp) != len(stat):
        print("Error")
        return
    
    min = math.inf
    ans = -1
    for i in range(len(sp)):
        if stat[i] == "1":
            continue
        diff = abs(sp[i] - n)
        if diff < min:
            min = diff
            ans = i
    if ans == -1:
        print("Not Found")
        return
    print(ans)


print("================Task4================")
func4([3, 1, 5, 4, 3, 2], "101000", 2)
func4([1, 0, 5, 1, 3], "10100", 4)
func4([4, 6, 5, 8], "1000", 4)

File: week2/assign2.js
function function1(name){
    // find out pos of the name
    // calculate the distance
    // check section and decide if it needs extra point
    // find the closest and farest
    
    const people = [
        {"name": "悟空", "pos": [0, 0]},
        {"name": "辛巴", "pos": [-3, 3]},
        {"name": "貝吉塔", "pos": [-4, -1]},
        {"name": "特南克斯", "pos": [1, -2]},
        {"name": "佛利沙", "pos": [4, -1]},
        {"name": "丁滿", "pos": [-1, 4]},
    ];
    
    const section1 = ["悟空", "辛巴", "貝吉塔", "特南克斯"];
    const section2 = ["佛利沙", "丁滿"];

    // find name itself and find pos
    const me = people.find(p => p.name === name);
    if (!me) return console.log("No exsist person");
    const POS = me.pos;
    
    // check section
    const check_section = (n) => (section1.includes(n) ? 1 : section2.includes(n) ? 2 : null);

    let target_section = check_section(name);

    // calculate the distance
    const extrapoint = 2;
    const new_list = [];
    for (let i = 0; i < people.length; i++){
        let person = people[i];
        // skip itself
        if (person.name === name){
            continue;
        }
        // distance 
        let distance_raw = Math.abs(person.pos[0] - POS[0]) + Math.abs(person.pos[1] - POS[1]);

        // different section
        const sec = check_section(person.name);
        if (sec !== target_section){
            distance_raw += extrapoint;
        }
        // make new list to put result
        new_list.push({[person.name] : distance_raw});
    }

    // find the closest
    let min = Infinity;
    let min_list = [];
    for (o of new_list){
        const [who, value] = Object.entries(o)[0]
        if (value < min){
            min = value;
            min_list = [who]; // don't use push. new value should overwrite everything.
        }
        else if (value === min){
            min_list.push(who);
        }
    }
    //find the farest
    let max = -Infinity;
    let max_list = [];
    for (o of new_list){
        const [who, value] = Object.entries(o)[0];
        if (value > max){
            max = value;
            max_list = [who];  // don't use push. new value should overwrite everything.
        }
        else if(value === max){
            max_list.push(who);
        }
    }
    console.log("最遠", max_list.join("、"), "; 最近", min_list.join("、"));
    
}

console.log("================Task1================")
function1("辛巴")
function1("悟空")
function1("佛利沙")
function1("特南克斯")


function func2(ss, start, end, criteria){
    // turn criteria into usable element
    // make the selector by criteria condition and put all qulify into result 
    // make the first one of result become best choice

    const[field, op , rawvalue] = criteria.split(/\s*(<=|>=|=)\s*/);
    let value;
    if (field === "r") value = Number(rawvalue);
    else if (field === "c") value = Number(rawvalue);
    else if (field === "name") value = rawvalue;
    else { console.log("Error"); return; }

    // make selector by condition
    let result = []
    for (let i = 0; i < ss.length; i++){
        const v = ss[i][field];
        // if condition is name 
        if (op === "="){
            if (field != "name"){
                return console.log("Error: name can only use '='");
            }
            if (v === value){
                result.push(ss[i]);
            }
        }
        // if condition is "r" or "c"
        if (op === ">="){
            if (Number(v) >= value){
                result.push(ss[i]);
            }
        }
        if (op === "<="){
            if (Number(v) <= value){
                result.push(ss[i]);
            }
        }
    }
    if (result.length === 0){ console.log("Not Found"); return; }
    // copy result and sort it
    let order = [...result];
    if (op === ">="){
        order.sort((a,b) => Number(a[field]) - Number(b[field])); // from small to big
    }
    else if (op === "<="){
        order.sort((a,b) => Number(b[field]) - Number(a[field])); // from big to small
    }
    const bset_choice = order[0];

    // put best choice into schedule
    // if same name in schedule check time avalible
    // if avalible then put best choice in and print out service
    function check_avalible(start, end, booking){
        if (start < booking.end && end > booking.start){
            return false
        }
        return true
    }

    // same service about to be used
    for (let i = 0; i < schedules.length; i++){
        if (bset_choice.name === schedules[i].name){
            if (check_avalible(start, end, schedules[i]) === false){
                console.log("Sorry");
                return
            }
        }
    }
    console.log(bset_choice.name);
    schedules.push({"name": bset_choice.name, "start": start, "end": end});
    return
    
}
    


const services = [
    {"name": "S1", "r": "4.5", "c":"1000"},
    {"name": "S2", "r": "3", "c":"1200"},
    {"name": "S3", "r": "3.8", "c":"800"},
];
schedules = [];
console.log("================Task2================")
func2(services, 15, 17, "c>=800");
func2(services, 11, 13, "r<=4");
func2(services, 10, 12, "name=S3");
func2(services, 15, 18, "r>=4.5");
func2(services, 16, 18, "r>=4");
func2(services, 13, 17, "name=S1");
func2(services, 8, 9, "c<=1500");


function func3(n){
    let list = [];
    list[0] = 25;

    let count1 = -2;
    let count2 = -3;
    let count3 = 1;
    let count4 = 2;

    for (let i = 1; i <= n; i++){
        if (i % 4 === 1){
            list[i] = list[i - 1] + count1;
        }
        else if (i % 4 === 2){
            list[i] = list[i - 1] + count2;
        }
        else if (i % 4 === 3){
            list[i] = list[i - 1] + count3;
        }
        else if (i % 4 === 0){
            list[i] = list[i - 1] + count4;
        }
    }
    console.log(list[n]);
}

console.log("================Task3================")
func3(1);
func3(5);
func3(10);
func3(30);

function func4(sp, stat, n){
    // put passages into car (take differences)
    // find out which car is unavalible and put the number to 100
    // find the fittest car (minimum)
    if (sp.length !== stat.length){
        return console.log("Error: sp and stat length not equal");
    }
    const len = sp.length

    let min = Infinity;
    let ans = -1;
    for (let i = 0; i < len; i++){
        if (stat[i] === "1") continue;
        let diff = Math.abs(sp[i] - n);
        if (diff <= min){
            min = diff;
            ans = i;
        }
    }
    if (ans === -1) return console.log("Error: Not found");
    console.log(ans);
}


console.log("================Task4================")
func4([3, 1, 5, 4, 3, 2], "101000", 2);
func4([1, 0, 5, 1, 3], "10100", 4);
func4([4, 6, 5, 8], "1000", 4);